### PR TITLE
Add slider for doubles

### PR DESCRIPTION
**Pull request by jhalmen**
for easier discussion about it

### CODE REVIEW

**Comment by jhalmen** [comment]
quite a few things broke with the new slider:
* slider text only updates after releasing mouse button
* cannot change slider by clicking to the left or right of current position 
* sliders start in their minimum position instead of central
* clicking any other widget type (pushbutton or switch) kills the application with 
    `AttributeError: 'MovablePushButton' object has no attribute 'value'`
* discrepancy between slider text and parameter value:
image

note: all of this worked before. i've tried to look into this myself, but i don't understand how these things interoperate and i don't find them self explanatory

File: pywisp/utils.py [hunk 1]
```diff
@@ -747,10 +747,72 @@ def updateData(self):
...
+
+    def setValue(self, value):
+        super().setValue(int((value - self._minValue) / self._valueRange * self._invStepSize))
+
+    def setMinimum(self, value):
+        if value > self._maxValue:
+            raise ValueError("Minimum limit cannot be higher than maximum")
+
+        self._minValue = value
+        super().setMinimum(int((self._minValue - self._minValue) / self._valueRange * self._invStepSize))
```
Comment (jhalmen) [inline pywisp/utils.py]: this is effectively `super().setMinimum(0)`

File: pywisp/utils.py [hunk 2]
```diff
@@ -747,10 +747,72 @@ def updateData(self):
...
+        super().setMaximum(int((self._maxValue - self._minValue) / self._valueRange * self._invStepSize))
+
+    def setMaximum(self, value):
+        if value < self._minValue:
+            raise ValueError("Minimum limit cannot be higher than maximum")
+
+        self._maxValue = value
+        super().setMinimum(int((self._minValue - self._minValue) / self._valueRange * self._invStepSize))
```
Comment (jhalmen) [inline pywisp/utils.py]: here too

File: pywisp/utils.py [hunk 3]
```diff
@@ -747,10 +747,72 @@ def updateData(self):
...
+        super().setMaximum(int((self._maxValue - self._minValue) / self._valueRange * self._invStepSize))
+
+    def setTickInterval(self, p_int):
+        self._stepSize = p_int
+
+        vRange = self._maxValue - self._minValue
+        self._invStepSize = vRange / self._stepSize
+
+        super().setMinimum(int((self._minValue - self._minValue) / self._valueRange * self._invStepSize))
```
Comment (jhalmen) [inline pywisp/utils.py]: and again.

might be helpful to just define a conversion function that you then use?

**Comment by jhalmen** [comment]
crash not fixed:
```Traceback (most recent call last):
  File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1495, in <lambda>
    widget.clicked.connect(lambda: self.remotePushButtonSendParameter(widget))
  File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1556, in remotePushButtonSendParameter
    self.remoteSliderUpdate(widget, value, sliderMoved=False)
  File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1595, in remoteSliderUpdate
    wid.label.setText(wid.widgetName + ': ' + "{:.3f}".format(widget.value))
AttributeError: 'MovablePushButton' object has no attribute 'value'
```

**Comment by TanDro** [comment]
> crash not fixed:
> 
> ```
>   File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1495, in <lambda>
>     widget.clicked.connect(lambda: self.remotePushButtonSendParameter(widget))
>   File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1556, in remotePushButtonSendParameter
>     self.remoteSliderUpdate(widget, value, sliderMoved=False)
>   File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1595, in remoteSliderUpdate
>     wid.label.setText(wid.widgetName + ': ' + "{:.3f}".format(widget.value))
> AttributeError: 'MovablePushButton' object has no attribute 'value'
> ```

Can't reproducible the crash.

**Comment by jhalmen** [comment]
> Can't reproducible the crash.

just opening a visualization with slider and pushbutton (e.g. MobRoSt) and clicking the pushbutton crashes it for me.

**Comment by jhalmen** [comment]
weird.. even trying to slide the slider now crashes with 
```
Traceback (most recent call last):
  File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1529, in <lambda>
    widget.sliderMoved.connect(lambda value: self.remoteSliderUpdate(widget, value))
  File "/home/halmenjo/git/pywisp/pywisp/gui.py", line 1598, in remoteSliderUpdate
    widget.label.setText(widget.widgetName + ': ' + "{:.3f}".format(widget.value))
AttributeError: 'MovableSlider' object has no attribute 'value'
```

**Comment by TanDro** [comment]
Please check if utils.py is up to date. The exception makes no sense^^

**Comment by jhalmen** [comment]
this would be my DoubleSlider. according to git it's the newest version
```
class DoubleSlider(QSlider):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        super().setMinimum(0)
        super().setMaximum(1)

        self._minValue = 0.0
        self._maxValue = 1.0

        self._stepSize = 1
        self._invStepSize = 1

    @property
    def _valueRange(self):
        vRange = self._maxValue - self._minValue
        self._invStepSize = vRange / self._stepSize

        return vRange

    @property
    def value(self):
        curValue = float(super().value())
        return curValue * self._stepSizet + self._minValue

    def setValue(self, value):
        super().setValue(int((value - self._minValue) / self._stepSize))

    def setMinimum(self, value):
        if value > self._maxValue:
            raise ValueError("Minimum limit cannot be higher than maximum")

        self._minValue = value
        super().setMinimum(0)
        super().setMaximum(int((self._maxValue - self._minValue) / self._stepSize))

    def setMaximum(self, value):
        if value < self._minValue:
            raise ValueError("Minimum limit cannot be higher than maximum")

        self._maxValue = value
        super().setMinimum(0)
        super().setMaximum(int((self._maxValue - self._minValue) / self._stepSize))

    def setTickInterval(self, p_int):
        self._stepSize = p_int

        vRange = self._maxValue - self._minValue
        self._invStepSize = vRange / self._stepSize

        super().setMinimum(0)
        super().setMaximum(int(self._invStepSize))

        super().setTickInterval(1)

    def minimum(self):
        return self._minValue

    def maximum(self):
        return self._maxValue
```

**Comment by jhalmen** [comment]
- [ ] widget Parameter scheint niciht richtig aus den `default.sreg` geladen zu werden
- [ ] slider text wird erst beim loslassen geupdated anstatt schon beim ziehen

**Comment by TanDro** [comment]
See last commit, that fix the last issues.

**Comment by jhalmen** [comment]
can confirm this fixes it.

nitpick: is it possible to initialize them to 0 ?

**Pull request merged**